### ISSUE TITLE
[build system] Link with shared lib (linux) without SONAME enabled (from autotools)

### ISSUE THREAD

**Issue by memsharded** [conan-io/conan#3230]
Creating packages with shared libs that doesn't have SONAME enabled is a problem for later consuming and linking with them:

- Check if there is some possible way to force autotools to link them
- Is it possible to detect it and warn? (Will be kind of "binary linters" feature, like https://github.com/conan-io/conan/issues/1115)
- At least, if not possible to develop something, document it in the docs.

**Comment by lasote** [comment]
Could you elaborate what is exactly the problem?

**Comment by memsharded** [comment]
It seems if you build a shared library in Linux without soname, then the linking with ``-lmything`` will fail to link with the ``libmything.so`` or ``libmything.so.1.2.3``. I think using the full filename to the linker might be a workaround.

**Comment by kepam** [comment]
Problem that we had was, that if we used recipe that was creating prebuild package with lib X.so that do not have soname, and later used it to link with Y.so (by building Y) then later when try to link Y to some Z lib/app paths to X was hard-coded, so if that was on other machine there was no way conan could override path to Z, so there was linker error about missing X.

Adding soname to X, solved this problem entirely.

**Comment by mduvanel** [comment]
As requested I am adding my detailed use-case for this problem here.

I am having an issue with a third-party library that happens to have no SONAME. Here is my scenario:

- I have repackaged this library (let's call it libexample.so) as a conan dependency, the way it is described here: https://docs.conan.io/en/latest/creating_packages/existing_binaries.html. Let's call this dependency examplepkg.
- I am developing a conan package which consists in an executable and a bunch of shared libraries, which depend on examplepkg. I am using the CMake generator
- I can build and link fine with this dependency, as well as run the application as long as I am on the same machine.
- Now if I try to run it inside a docker container (using conan install), it tries to load the library from a non-existing path in /home/vagrant/.conan, which is the path where the example library is located on the machine that performed the build

Tracing the issue back I noticed that libexample.so has no SONAME field in its ELF header, and that the other shared libraries which are linking with it have an absolute path in the NEEDED field:

```
  NEEDED               libz.so.1
  NEEDED               /home/vagrant/.conan/data/examplepkg/1.0/channel/stable/package/de9c231f84c85def9df09875e1785a1319fa8cb6/lib/libexample.so
  NEEDED               libxerces-c-3.1.so
```

To try to fix this problem I made an attempt to set the IMPORTED_NO_SONAME property for that library in my CMakeLists.txt as follows:

```
    include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
    conan_basic_setup(TARGETS)
    set_property(TARGET CONAN_LIB::examplepkg_example PROPERTY IMPORTED_NO_SONAME 1)
```

However I see no change in the behaviour. But the wording in the CMake documentation is a bit vague (emphasis mine):

> CMake **may** adjust generated link commands **for some platforms** to prevent the linker from using the path to the library in place of its missing soname. Ignored for non-imported targets.

I realize this might be more a CMake integration issue, however I am probably not the first one to encounter this issue with conan and I wonder if there are known workarounds for this type of improperly packaged libraries when using CMake?

**Comment by mduvanel** [comment]
I played a bit with the link.txt generated by cmake for my consuming library that links against libexample.so. Here are my findings:
- the options that are generated by default look like this:
`{snip} -Wl,-rpath,/absolute/path/to/conan/cache/example/lib /absolute/path/to/conan/cache/example/lib/libexample.so` which works fine if libexample.so has a SONAME (linked with the SONAME only), but not if it has no SONAME (linked with the absolute path).
- I tried to manually patch the link.txt file by adding the library search path explicitly like this, without success:
`{snip} -L/absolute/path/to/conan/cache/example/lib -Wl,-rpath,/absolute/path/to/conan/cache/example/lib /absolute/path/to/conan/cache/first/lib/libexample.so`
- I then switched the library name to the absolute name, and there it worked:
`{snip} -L/absolute/path/to/conan/cache/example/lib -Wl,-rpath,/absolute/path/to/conan/cache/example/lib -lexample`
I now have a shared library linked with just the non-existing SONAME:

```
  NEEDED               libz.so.1
  NEEDED               libexample.so
  NEEDED               libxerces-c-3.1.so
```

However I have no direct control over how the targets are setup, and just changing the dependency name in target_link_libraries() from `CONAN_PKG::examplepkg` to `example` does not work out of the box, because then I also lose all the inherited include folders from examplepkg.

I have the feeling that a possible solution to handle this use-case could be the conan_basic_setup() function, where an extra option could be passed to change how the library is consumed in order to provide proper linking.

On the other hand my use-case might be a special case in the sense that my example library follows the basic libXXX.so naming convention, I am not sure this also works if the library is just named example.so.

Let me know if I can help further this discussion!

**Comment by sourcedelica** [comment]
The problem is that `conan_package_library_targets` creates the `CONAN_LIB` libraries as `UNKNOWN IMPORTED`.  If `SHARED IMPORTED` is used along with `IMPORTED_NO_SONAME` instead then CMake will link using `-L ${CONAN_LIB_DIRS_FOO} -lfoo` and the `DT_NEEDED` entry will not include the full path.

See https://gitlab.kitware.com/cmake/cmake/issues/18052.

Conan needs a way to differentiate shared from static libraries in `cpp_info`.  That way it can use that information to add libraries as `SHARED IMPORTED` and `STATIC IMPORTED` instead of `UNKNOWN IMPORTED`.

**Comment by lasote** [comment]
Thanks for the info @sourcedelica. @mduvanel If you could verify that it works for you it would be nice.  
@memsharded do you have some thoughts about propagating a "share" condition in the `cppinfo` object? Is it maybe something we might be able to use for other concerns in the graph?

**Comment by ericriff** [comment]
I just came across this myself. I'm packaging a prebuilt shared library with no soname, the test package builds fine but if I run `ldd` on it I see that is using absolute paths to find the library.
Then, if I consume the library I have issues linking.
I would like to know if you guys figured something out.

**Comment by kepam** [comment]
Mine solution (while back) was to hack prebuild library by adding SONAME.

**Comment by ericriff** [comment]
I was looking into the same approach.
I'm using `patchelf` to add a `SONAME` to the library.

**Comment by mduvanel** [comment]
The solution we put in place at the time was different, but more complicated.

Instead of altering the problematic library itself, we hacked the consuming library's build process so that it would edit the link.txt files produced by CMake before building, to replace the `/absolute/path/to/conan/cache/first/lib/libexample.so` parts into `-lexample`.

Practically, it meant doing something similar to this for the *consuming* library:

```
def patch_library_name(name):
    # for each 'link.txt' file:
    #     s/regexp_for_path_to_libname.so/-lname/

def build(self):
    cmake = CMake(self)
    patch_library_name('example')
    cmake.build()
```

Sorry for not being able to share actual code, I don't have access to it anymore.

Anyway, in hindsight it seems more cumbersome than to edit the header of the prebuilt library to add a `SONAME`, in particular because our solution does not scale well : it requires doing the same patching for all the consuming packages. We only had one so we didn't bother.

**Comment by lasote** [comment]
We are going to try to take a look at this but still not sure about the best approach.

**Comment by atrelinski** [comment]
There is one more business case to consider.

If the library is provided without the soname and it is signed, You actually are in the dead-end position when this library is consumed by several other libraries.
Case which I am talking about is when You are trying to migrate to Conan not a particular library but whole tree dependent libraries, where one of them must be consumed as binary (license or paid support agreement) and can’t be changed ([patchelf](https://github.com/NixOS/patchelf)) since it is signed.
Also You can’t patch all recipes that consume that library since you don’t control them.

**Comment by memsharded** [comment]
> Also You can’t patch all recipes that consume that library since you don’t control them.

It seems that for this very problematic use case, of a vendor of closed source signed binaries, it is very likely that users will actually control the consuming recipes.

@sourcedelica suggestion to add "shared" specification: https://gitlab.kitware.com/cmake/cmake/-/issues/18052, is related exclusively to CMake, while at least the original issue was reported for Autotools. it might still help for some use cases, but I am thinking how this would make recipes even more complex for this edge-case, apparently so far it has not been a big issue to generate UNKNOWN targets (it is common also for upstream CMake find modules to do this).

Hacking the consumer CMake link.txt is also very specific to CMake, and seems too much of a hack to implement.

**Comment by franramirez688** [comment]
As far as I have been figuring out this issue, I think that the "easiest" way to solve the problem would be to use [patchelf](https://github.com/NixOS/patchelf) (as @ericriff and @kepam explained above) as a way to detect/set the SONAME tag:

* Check if SONAME exists: `patchelf --print-soname mysharedlib.so.1.2.3`
* Set a SONAME name: `patchelf --set-soname mysharedlib.so.1 mysharedlib.so.1.2.3`

Autotools does not have any way to detect it even using directly [libtool](https://www.gnu.org/software/libtool/) which is in charge of managing the creation of those shared libraries. At least, I was not able to find out anything regarding it.

By the way, Autotools and CMake are applying the SONAME flag by default so it should not be a typical issue.

**Comment by atrelinski** [comment]
I also believe `patchelf` is the easiest way, however this is not the point.
I wonder how many other developers face this problem.

Our case is that binary is signed, so You can't change it.
Yes, we have access to sources (this is not a closed source library), but process of their compilation is complex and require machine that have several things installed and it does not base on cmake. More over even if we successfully build the library, I don't think we will be able to sign it. l That is why we are consuming binaries.
We are trying to shift tree of libraries to Conan, not one library, and this SONAME missing library is in its middle.

**Comment by franramirez688** [comment]
Hi @atrelinski 

I'm trying to reproduce the issue in my system and I was not able to achieve it. Let me explain my case to check together if I'm doing something wrong or perhaps, it could be solved for the latest Conan version.

### Environment Details (include every applicable attribute)
* Operating System+version: macOS 11.6
* Compiler+version: Clang 12.0.0
* Conan version: 1.41
* Python version: 3.9.7

### Steps to reproduce
Create an empty project with the newest CMake template from Conan:
```
conan new nosoname/1.0 -m cmake_lib
```
Now, let's modify something to create a shared lib without SONAME and consume it through the default test.

**./CMakeLists.txt**

```cmake
# Adding NO_SONAME flag to main library
set_target_properties(nosoname PROPERTIES PUBLIC_HEADER "src/nosoname.h" NO_SONAME 1)
```
**./conanfile.py**
```python
class NosonameConan(ConanFile):
    name = "nosoname"
    version = "1.0"
    # ....

    # Enabling shared 
    default_options = {"shared": True, "fPIC": True}
```
Create the package:
```
conan create .
```
Everything is built/linked as expected:
```
➜  $ conan create . 
....

[ 50%] Building CXX object CMakeFiles/nosoname.dir/src/nosoname.cpp.o
[100%] Linking CXX shared library libnosoname.dylib
[100%] Built target nosoname

...

[ 50%] Building CXX object CMakeFiles/example.dir/src/example.cpp.o
[100%] Linking CXX executable example
[100%] Built target example
nosoname/1.0 (test package): Running test()
Capturing current environment in deactivate_conanrunenv-release-x86_64.sh
Configuring environment variables
Configuring environment variables
nosoname/1.0: Hello World Release!
  nosoname/1.0: __x86_64__ defined
  nosoname/1.0: __cplusplus199711
  nosoname/1.0: __GNUC__4
  nosoname/1.0: __GNUC_MINOR__2
  nosoname/1.0: __clang_major__12
  nosoname/1.0: __apple_build_version__12000032
```
To ensure that the `dylib` was correctly built without the SONAME:
```
➜  $ objdump -p ~/.conan/data/nosoname/1.0/_/_/package/ba3872b419f0b5f5948b43bc6817447b031fa997/lib/libnosoname.dylib | grep SONAME
➜  $
```
Am I missing anything?

I know that your case is not a CMake-related one but I am wondering if you could provide an example to try to reproduce locally the issue. A simple example like mine above would be enough.

**Comment by atrelinski** [comment]
Hi @franramirez688 

My path is different.
You need 3 components: 'LibNoSoname' -> 'OtherLib' -> 'Executable' to see the problem.
The 'LibNoSoname' is a Conan package consumed by the 'OtherLib' and the 'OtherLib' is consumed as a Conan package by 'Executable'.
I see @kepam reported the same path to reproduce problem [link](https://github.com/conan-io/conan/issues/3230#issuecomment-406328481) and @mduvanel [link](https://github.com/conan-io/conan/issues/3230#issuecomment-435930192) as well.

PS: Long time ago, I've tested in my system and for path 'LibNoSoname' -> 'Executable` everything was working as expected on CentOS.

**Comment by franramirez688** [comment]
Hi @atrelinski 

I have tried it with that path as you can see here: https://github.com/franramirez688/reproduce-bugs/tree/bug/soname
It's working fine as well so I don't know if there is another condition that I'm missing or perhaps it could be already fixed using the new `conan.tools.cmake` classes like `CMakeDeps`, `CMakeToolchain` and `CMake`.
I'd like to prepare something with Autotools too to see what happens.

**Comment by atrelinski** [comment]
@franramirez688 

Wow! Very nice professional approach with the example!
Thank You!

My example is complex and I had troubles with simplifying it.
I am using the `cmake_find_package_multi` generator, please could You try with it?

**Comment by franramirez688** [comment]
Thanks @atrelinski 
I just added another branch trying the same example but using older generators like `cmake_find_package_multi`.
You can try it here: https://github.com/franramirez688/reproduce-bugs/tree/bug/soname-legacy
It's executing fine the application as well.

**Comment by atrelinski** [comment]
Hi @franramirez688 

I have bad news.
I've done experiment and I was unable to reproduce problem any longer if it is outside docker environment, with default conan paths, with conan 1.42.0 on CentOS.
I strongly believe problem is still there since I am observing the same warnings as before, but since we've changed environment, it would take me a lot of time to reproduce the problem. And even if I would do that and problem is still there, there are too many factors involved (docker environment, changing default conan paths, conan version, OS and probably more) that I won't be able to separate problem in clear conditions, so You can reproduce it and fix it.

Bottom line,
I must withdraw my arguments (path that leads to bug), since on conan 1.42.0 on CentOS it is working correctly for us.

I would like to thank You a lot for Your involvement, problem reproducing effort and help!

**Comment by franramirez688** [comment]
Hi @atrelinski 

Thank you so much for your comment and I agree, I think the problem could be still there but it'll be likely a corner case that we're not able to reproduce for now.
By the way, I have opened a little PR (https://github.com/conan-io/conan/pull/10107) to simulate these two cases at least for Linux so I'll close this issue for now when it's merged.
For sure, it'll be reopened if any other problem appears again.
Thanks again.

**Comment by fanselm** [comment]
Hello, just here saying the problem is still there.

In @franramirez688 example (https://github.com/franramirez688/reproduce-bugs/tree/bug/soname-legacy) you don't see it as when you run the executable the linker can still find the shared libraries through the absolute paths. Now if you instead copy `libnosoname.so` and `liblibB.so` to a separate local folder and then remove them from the conan cache, set LD_LIBRARY_PATH to point to the local folder and then run the program then you get a linker error: `ld can't find /abs/path/to/.conan/data/nosoname/1.0.0/_/_/package/xxx/lib/libnosoname.so`.

As mentioned above: it is a CMake problem. For libraries with no SONAME one has to set the property [IMPORTED_NO_SONAME](https://cmake.org/cmake/help/v3.22/prop_tgt/IMPORTED_NO_SONAME.html) which will ONLY work if the target has been explicitly defined as SHARED - not UNKNOWN. So the `cmake_find_package` or `CMakeDeps` generated CMake macros must somehow figure out if the library is shared (from Conan options OR it could do so from the file extension, as this is a Linux-only issue as far as I know) and whether it has an SONAME or not.

It might be that a proper general solution is too complicated for such fringe behavior.

It is still unknown why some library vendors don't provide an SONAME for their prebuilt libraries. In my case it is Intel MKL 2017 libraries that doesn't do it. In never versions of Intel oneAPI they have changed this - they likely learned from their own mistakes. For now I'm just doing like everyone else: patchelf the MKL libraries and hope that it doesn't break anything

**Comment by franramirez688** [comment]
Hi @fanselm 

Thanks a lot for your feedback, it's really appreciated.
I'm going to perform another test to try to reproduce exactly what you're saying in my local so we could have anything failing on our side and try to figure out a possible solution.
I'll update this thread with the result and reopen the issue in that case as well.

**Comment by fanselm** [comment]
I have now found more libraries that also don't have an SONAME :(

Now I have changed my mind about the cmake generators: CMake should never link in shared libraries with the full path. That is not the "Linux-way", the whole GNU environment assumes that you will link with `-L/path/to/lib -lname` with the inherent risk of picking up libraries from some place else if you're not careful enough. That means one should should not use `find_library()` to actually find the library paths. Conan also doesn't need that - it should already know where the files are located: in package/xxx/lib, so I don't know why these extra steps are needed. If CMake in general think `find_library()` is a good idea it should have some functionality that just extracts the directory and adds that to `target_link_directories()`.

From a package manager developers point of view it is of course easier and safer to pass the absolute paths to libraries to the linker - that way you ensure that you link the exact libraries you want and not some that are randomly found other places through wrong search order. But `ld` and the GNU ecosystem was apparently not designed for this, as they assume you are always root and only install packages through your distributions package manager directly into `/usr/lib` and god forbid that you have multiple versions installed. Anyway it is easier to change the cmake generator than trying the change the `ld` behavior and GNU philosophy.

**Comment by fanselm** [comment]
Just adding: The default behavior of ld (or as invoked through gcc or g++) is to NOT set an SONAME. This means that there may be quite a lot of libraries out there with this problem. CMake will typically invoke ld setting SONAME, but there's still a lot of libraries that do not use CMake - especially in the academic scientific community.

My solution is to either use `patchelf` or to provide custom `<PKG>Config.cmake` files.

However, if you want Conan to automatically generate such CMake config files Conan needs to keep track of whether the library files set SONAME or not. This may seem like unnecessary information to carry around - in that case Conan can try to talk to KitWare about changing the behavior of CMake as it is in general a problem of using targets with IMPORTED_LOCATION. Maybe CMake could include an internal tools that actually reads the ELF meta data and checks whether SONAME is set for imported files.

What do you say? At least an issue could be files for CMake to let Kitware know of the problem.

**Comment by atrelinski** [comment]
Great, Great, Great! job @fanselm !!

@franramirez688 @memsharded 
Can we open this issue since we understood problem can reproduce the problem?

**Comment by memsharded** [comment]
Closed by https://github.com/conan-io/conan/pull/11318, will be in next 2.0-beta.1!

**Comment by atrelinski** [comment]
Thank You for understanding!
